cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/ConsumerWorkingSetOperations.java

```diff
@@ -0,0 +1,193 @@
+      if (consumablePartition.incrementNumFailures() < consumerConfiguration.getNumRetries()) {
+        consumablePartition.resetState();
+      } else {
+        discardedPartitions.add(key);
+        workingSet.remove(key);
+      }
+    }
+    if (!discardedPartitions.isEmpty()) {
+      LOG.warn("Discarded keys due to being retried {} times: {}",
+               consumerConfiguration.getNumRetries(), discardedPartitions);
+    }
+  }
+
+  // ensure that user doesn't try to commit/abort a partition that isn't in progress
+  private void assertInProgress(ConsumablePartition consumablePartition) {
+    if (!consumablePartition.isInProgress()) {
+      throw new IllegalStateException(String.format("Partition not in progress: %s",
+                                                    consumablePartition.getPartitionKey()));
+    }
+  }
+
+  // ensure that user doesn't try to commit/abort a partition that is expired
+  private void assertNotExpired(ConsumablePartition consumablePartition, long expiryBorder) {
+    if (consumablePartition.getTimestamp() < expiryBorder) {
+      throw new IllegalStateException(String.format("Attempting to operate on an expired partition: %s",
+                                                    consumablePartition.getPartitionKey()));
+    }
+  }
+
+  // returns a timestamp which determines partition expiry. Partitions with a timestamp smaller (older) than this value
+  // are considered 'expired'.
+  private long getExpiryBorder() {
+    long now = System.currentTimeMillis();
+    long expirationTimeoutMillis = TimeUnit.SECONDS.toMillis(consumerConfiguration.getTimeout());
+    return now - expirationTimeoutMillis;
+  }
+
+  // Goes through all partitions. If any IN_PROGRESS partition is older than the configured timeout, reset its state
+  // to NEW, unless it has already been retries the configured number of times, in which case it is discarded.
+  private void doExpiry() {
+    long expiryTime = getExpiryBorder();
+    List<PartitionKey> expiredPartitions = new ArrayList<>();
+    List<PartitionKey> discardedPartitions = new ArrayList<>();
+    for (ConsumablePartition partition : workingSet.getPartitions()) {
+      if (partition.isInProgress() && partition.getTimestamp() < expiryTime) {
+        // either reset its processState, or remove it from the workingSet, depending on how many tries it already has
+        if (partition.incrementNumFailures() < consumerConfiguration.getNumRetries()) {
+          partition.resetState();
+        } else {
+          workingSet.remove(partition.getPartitionKey());
+        }
+        expiredPartitions.add(partition.getPartitionKey());
+      }
+    }
+    if (!expiredPartitions.isEmpty()) {
+      LOG.warn("Expiring in progress partitions: {}", expiredPartitions);
+      if (!discardedPartitions.isEmpty()) {
+        LOG.warn("Discarded keys due to being retried {} times: {}",
+                 consumerConfiguration.getNumRetries(), discardedPartitions);
+      }
```

Comment:
Updated as a commit in this PR: https://github.com/caskdata/cdap/commit/de8d10a9c168700089f560ef0cd3ef7ebc92ae04